src/message.rs

```diff
@@ -36,14 +36,160 @@ impl Deserialize for Version {
+        match self {
+            Params::Positional(xs) => Value::Array(xs),
+            Params::Named(x) => Value::Object(x),
+        }
+    }
+
+    /// Force the params to be represented in positional form
+    pub fn into_positional(self) -> Self {
+        match self {
+            Params::Named(x) => {
+                let values = x.into_iter()
+                    .map(|(_, v)| v)
+                    .collect();
+                Params::Positional(values)
+            },
+            x => x,
+        }
+    }
+}
+
+// Helper fuction for skipping the params field serialization when there are no params
+pub fn params_empty(x: &Option<Params>) -> bool {
+    match *x {
+        Some(ref x) => x.is_empty(),
+        None => true,
+    }
+}
+
+// TODO: how do we sanitize uses of `json_internal!()`? check the rocket crate
+// TODO: missing docs
+#[macro_export]
+macro_rules! params {
+
+    //////////////////////////////////////////////////////////////////////////
+    // The main implementation.
+    //
+    // Must be invoked as: params!($($json)+)
+    //////////////////////////////////////////////////////////////////////////
+
+    () => {
+        None
+    };
+
+    // Review: do we want to use this to send a `"params": null` as opposed to skipping it?
+    //         params!() or None can be the skip while params!(null) will actually send null
+    //         Notes:
+    //           - Params::Null needs to be added
+    //           - `None` and `Some(Params::Null)` are "almost" equivalent on server side
+    (null) => {
+        None
+    };
+
+    (true) => {
+        Some($crate::Params::Positional(vec![$crate::macro_exports::Value::Bool(true)]))
+    };
+
+    (false) => {
+        Some($crate::Params::Positional(vec![$crate::macro_exports::Value::Bool(false)]))
+    };
+
+    ([]) => {
+        {
+            Some($crate::Params::Positional(vec![]))
+        }
+    };
+
+    ([ $($tt:tt)+ ]) => {
+        {
+            match $crate::macro_exports::Value::Array(json_internal!(@array [] $($tt)+)) {
+                $crate::macro_exports::Value::Array(xs) => Some($crate::Params::Positional(xs)),
+                _ => unreachable!(),
+            }
+        }
+    };
+
+    ({}) => {
+        {
+            Some($crate::Params::Named($crate::Map::new()))
+        }
+    };
+
+    ({ $($tt:tt)+ }) => {
+        {
+            let mut object = $crate::macro_exports::Map::new();
+            json_internal!(@object object () $($tt)+);
+            Some($crate::Params::Named(object))
+        }
+    };
+
+    // Force structs to be serialized as positional
+    (pos $other:expr) => {
+        match $crate::macro_exports::to_value(&$other).unwrap() {
+            // Turn the object into positional discarding the names
+            $crate::macro_exports::Value::Object(x) => {
+                let mut xs = Vec::with_capacity(x.len());
+                for (_,v) in x {
+                    xs.push(v);
+                }
```

Comment:
Personally, I prefer something like `x.map(|(_, v)| v).collect()`. I even believe this might be more idiomatic. But maybe that's just a personal preference, so just pointing out the possibility.